lib/src/consumer.dart

```diff
@@ -1,64 +1,104 @@
 import 'package:flutter/widgets.dart';
 import 'package:provider/src/provider.dart';
 
+/// {@template provider.consumer}
 /// Obtain [Provider<T>] from its ancestors and pass its value to [builder].
 ///
 /// [builder] must not be null and may be called multiple times (such as when provided value change).
+///
+/// ## Performances optimizations:
+///
+/// {@macro provider.consumer.child}
+/// {@endtemplate}
 class Consumer<T> extends StatelessWidget {
+  /// {@template provider.consumer.constructor}
+  /// Consumes a [Provider<T>]
+  /// {@endtemplate}
+  Consumer({
+    Key key,
+    @required this.builder,
+    this.child,
+  })  : assert(builder != null),
+        super(key: key);
+
+  // fork of the documentation from https://docs.flutter.io/flutter/widgets/AnimatedBuilder/child.html
+  /// The child widget to pass to [builder].
+  /// {@template provider.consumer.child}
+  ///
+  /// If a builder callback's return value contains a subtree that does not depend on the provided value,
+  /// it's more efficient to build that subtree once instead of rebuilding it on every change of the provided value.
+  ///
+  /// If the pre-built subtree is passed as the child parameter, [Consumer] will pass it back to the builder function so that it can be incorporated into the build.
+  ///
+  /// Using this pre-built child is entirely optional, but can improve performance significantly in some cases and is therefore a good practice.
+  /// {@endtemplate}
+  final Widget child;
+
+  /// {@template provider.consumer.builder}
   /// Build a widget tree based on the value from a [Provider<T>].
   ///
   /// Must not be null.
-  final Widget Function(BuildContext context, T value) builder;
-
-  /// Consumes a [Provider<T>]
-  Consumer({Key key, @required this.builder})
-      : assert(builder != null),
-        super(key: key);
+  /// {@endtemplate}
+  final Widget Function(BuildContext context, T value, Widget child) builder;
```

Comment:
Any reason for this particular sorting? ScopedModel has [context, child, value](https://pub.dartlang.org/documentation/scoped_model/latest/scoped_model/ScopedModelDescendantBuilder.html), so having the same ordering would make some people's life easier. Not just when upgrading from one to another, but also when developing — I personally have muscle memory for context, child, value. 

It also corresponds better with Flutter's [TransitionBuilder](https://docs.flutter.io/flutter/widgets/TransitionBuilder.html).

On the other hand, [ValueWidgetBuilder](https://docs.flutter.io/flutter/widgets/ValueWidgetBuilder.html) has the same ordering as you, and it's the closest to `Consumer` that we already have in Flutter. It's just less known than ScopedModel.

I don't have any preference. "context, child, value" for historic purposes, "context, value, child" for more consistency with the framework itself. Make your pick. Just making sure we're intentional about this.

Reply:
It follows the "optional parameters are placed after required parameters" rule.
While the typing system of Dart doesn't support it yet, technically `child` is optional.